### DISCUSSION TITLE
/bin/sh: [npm,: not found. ( Kubernetes )

### DISCUSSION THREAD

**Discussion by tron4x** [vuejs/vitepress#4159] [Q&A] [answered]
Hello,

have some problem running vitepress on kubernetes.

My dockerfile looks like this:

```
FROM node:lts-alpine AS base

FROM base AS buildtime
RUN apk add --no-cache libc6-compat git
RUN apk update && apk add bash && apk add npm

# install dependencies
FROM buildtime AS deps
WORKDIR /app
COPY package.json package-lock.json ./

RUN npm ci

FROM buildtime AS builder
RUN apk add --no-cache libc6-compat git
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .
RUN npm add -D vitepress
RUN npm fund

CMD ["npm", "run", "docs:dev"]
````

I have created a simple helm chart to deploy vitepress on kubernetes.
Can be found here: [Helm Chart](https://github.com/tron4x/docs-helm)
Everything is starting with the exception of the pod.

In ArgoCD all resources are up running. The pod not.
Logs  of pod:

```
/bin/sh: [npm,: not found
```

I don't understand why npm is not found even though I install it explicitly.
Everything works locally on my Mac.

I've been looking for solutions for a while, but unfortunately I haven't found anything yet.
Can you help me here please?

Thank you

**Comment by brc-dd** [comment]
Not sure, gpt solution:

```
...

FROM builder
CMD ["npm", "run", "docs:dev"]
```

Alternatively, ditch npm, just run `./node_modules/.bin/vitepress dev docs` (or whatever your command is).

**Comment by tron4x** [comment]
It doesn't work.
Have simplify my dockerfile witn an older node version, but the result is the same.
```
FROM node:18-alpine3.20 AS base
RUN apk add --no-cache git

WORKDIR /app
COPY package.json package-lock.json ./
RUN node -v && npm -v

RUN npm ci

COPY . .

RUN npm install -D vitepress
RUN npm i vitepress

RUN ls -la /app/node_modules/.bin/vitepress

EXPOSE 5137

CMD ["./app/node_modules/.bin/vitepress"]
```

`Error:  ./app/node_modules/.bin/vitepress not found`

With dot or without, it doesn't makes matter.

You could also use nginx as a web server and display the static pages but I think it looks better to use vitepress or npm.
Has anyone got this to work on Kubernetes ?

**Reply by brc-dd** [reply]
> Has anyone got this to work on Kubernetes?

Most likely it is not related to vitepress. That binaries are missing somehow. Someone who has worked more on docker/k8s might be able to comment. Ask on StackOverflow or something. Try running any command from npm, it should fail similarly.

Also, I doubt that many people are using k8s with vitepress because you simply don't need it. You shouldn't be deploying the dev version anyway. It will be too slow. And `vitepress build` generates static site, which can be served with just nginx.

**Comment by tron4x** [comment]
You mean this here:  [Nginx](https://vitepress.dev/guide/deploy#nginx)
That's how it works.
The second Docker image I built also has all the nginx binaries.
The nginx conf file is provided in the pod via a config map.
I'll post the finished deployment here. I still have to test it a bit.

I was hoping there was another way.
I'll try again, maybe I'll think of something else.

Thanks for any tips though.

**Comment by tron4x** [comment, marked as answer]
Hi,

now its working on Kubernetes ( with nginx ). 😄 
I spent a while on this until it worked properly.
This Dockerfile do the job:

```
FROM node:alpine3.20 AS base
RUN apk add --no-cache libc6-compat git

# Install build dependencies
FROM base AS deps
WORKDIR /app
COPY package.json package-lock.json ./
RUN npm ci

# Build static website
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .
RUN npm add -D vitepress
RUN npm run docs:build

# Use nginx to serve the static website
FROM nginx:stable-alpine AS final
WORKDIR /usr/share/nginx/html
COPY --from=builder /app/.vitepress/dist ./
COPY default.conf /etc/nginx/conf.d/default.conf

# Expose necessary ports
EXPOSE 8080 8443

# Start nginx
CMD nginx -g 'daemon off;'
```

default.conf is from [here](https://vitepress.dev/guide/deploy#nginx)
root location needs to be adjusted. In my case:` /usr/share/nginx/html`

Kubernetes resources are up and running:
```
NAME                             READY   STATUS    RESTARTS   AGE
pod/home-docs-5c68b994c9-smsgj   1/1     Running   0          15m

NAME                TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
service/home-docs   ClusterIP   10.43.70.150   <none>        80/TCP    6h28m

NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/home-docs   1/1     1            1           6h28m

NAME                                   DESIRED   CURRENT   READY   AGE
replicaset.apps/home-docs-5c68b994c9   1         1         1       15m
replicaset.apps/home-docs-7874977488   0         0         0       6h28m
replicaset.apps/home-docs-dd8bcfcc6    0         0         0       6h25m
replicaset.apps/home-docs-ddb8dffc8    0         0         0       6h24m
```

Logs from pod:
```
> k logs pod/home-docs-5c68b994c9-smsgj -n home

2024/08/25 08:33:15 [notice] 1#1: using the "epoll" event method
2024/08/25 08:33:15 [notice] 1#1: nginx/1.26.2
2024/08/25 08:33:15 [notice] 1#1: built by gcc 13.2.1 20240309 (Alpine 13.2.1_git20240309)
2024/08/25 08:33:15 [notice] 1#1: OS: Linux 6.1.0-22-amd64
2024/08/25 08:33:15 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2024/08/25 08:33:15 [notice] 1#1: start worker processes
2024/08/25 08:33:15 [notice] 1#1: start worker process 7
2024/08/25 08:33:15 [notice] 1#1: start worker process 8
2024/08/25 08:33:15 [notice] 1#1: start worker process 9
2024/08/25 08:33:15 [notice] 1#1: start worker process 10
2024/08/25 08:33:15 [notice] 1#1: start worker process 11
2024/08/25 08:33:15 [notice] 1#1: start worker process 12
2024/08/25 08:33:15 [notice] 1#1: start worker process 13
2024/08/25 08:33:15 [notice] 1#1: start worker process 14
2024/08/25 08:33:15 [notice] 1#1: start worker process 15
2024/08/25 08:33:15 [notice] 1#1: start worker process 16
2024/08/25 08:33:15 [notice] 1#1: start worker process 17
2024/08/25 08:33:15 [notice] 1#1: start worker process 18
2024/08/25 08:33:15 [notice] 1#1: start worker process 19
2024/08/25 08:33:15 [notice] 1#1: start worker process 20
2024/08/25 08:33:15 [notice] 1#1: start worker process 21
2024/08/25 08:33:15 [notice] 1#1: start worker process 22
78.46.98.221 - - [25/Aug/2024:08:33:16 +0000] "GET / HTTP/1.1" 200 10655 "-" "kube-probe/1.28" "-"
78.46.98.221 - - [25/Aug/2024:08:33:22 +0000] "GET / HTTP/1.1" 200 10655 "-" "kube-probe/1.28" "-"
78.46.98.221 - - [25/Aug/2024:08:33:22 +0000] "GET / HTTP/1.1" 200 10655 "-" "kube-probe/1.28" "-"
```

The trick here is to set the command for starting nginx correctly.
The usual would be: `CMD ["nginx", "-g", "daemon off;"]`
But that doesn't work! I still get: `["nginx" not found.` on Kubernetes.

Passing CMD as a shell argument, however, works very well!   👍
